csrc/multidevice/communicator.cpp

```diff
@@ -0,0 +1,143 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#ifdef USE_DISTRIBUTED
+
+#include <multidevice/communicator.h>
+#ifdef USE_C10D_GLOO
+#include <torch/csrc/distributed/c10d/ProcessGroupGloo.hpp>
+#endif
+#ifdef USE_C10D_NCCL
+#include <torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp>
+#endif
+#ifdef USE_C10D_UCC
+#include <torch/csrc/distributed/c10d/ProcessGroupUCC.hpp>
+#endif
+
+namespace nvfuser {
+
+// Parse the environment to retrieve MPI rank and MPI world size and sets rank
+// and size accordingly Returns 0 in case of success, 1 otherwise
+int parseEnv(
+    RankType& rank,
+    int64_t& size,
+    RankType& local_rank,
+    int64_t& local_size) {
+  char* env = nullptr;
+
+  // retrieves the rank of the current process
+  env = std::getenv("OMPI_COMM_WORLD_RANK");
+  if (!env) {
+    env = std::getenv("WORLD_RANK");
+    if (!env) {
+      return 1;
+    }
+  }
+  rank = std::atoi(env);
+
+  // retrieves the size of the communicator
+  env = std::getenv("OMPI_COMM_WORLD_SIZE");
+  if (!env) {
+    env = std::getenv("WORLD_SIZE");
+    if (!env) {
+      return 1;
+    }
+  }
+  size = std::atoi(env);
+
+  // retrieves the size of the communicator
+  env = std::getenv("OMPI_COMM_WORLD_LOCAL_RANK");
+  if (!env) {
+    env = std::getenv("WORLD_LOCAL_RANK");
+    if (!env) {
+      return 1;
+    }
+  }
+  local_rank = std::atoi(env);
+
+  // retrieves the size of the communicator
+  env = std::getenv("OMPI_COMM_WORLD_LOCAL_SIZE");
+  if (!env) {
+    env = std::getenv("WORLD_LOCAL_SIZE");
+    if (!env) {
+      return 1;
+    }
+  }
+  local_size = std::atoi(env);
+
+  return 0;
+}
+
+// creates and return a process group backend
+c10::intrusive_ptr<c10d::Backend> createBackend(
+    CommunicatorBackend backend,
+    ::c10::intrusive_ptr<c10d::TCPStore> store,
+    RankType rank,
+    int64_t size) {
+#ifdef USE_C10D_NCCL
+  if (backend == CommunicatorBackend::nccl) {
+    auto pg_opts = c10::make_intrusive<::c10d::ProcessGroupNCCL::Options>();
+    return c10::make_intrusive<::c10d::ProcessGroupNCCL>(
+        store, rank, size, pg_opts);
+  }
+#endif
+
+#ifdef USE_C10D_GLOO
+  if (backend == CommunicatorBackend::gloo) {
+    auto pg_opts = c10d::ProcessGroupGloo::Options::create();
+    return c10::make_intrusive<::c10d::ProcessGroupGloo>(
+        store, rank, size, pg_opts);
+  }
+#endif
+
+#ifdef USE_C10D_UCC
+  if (backend == CommunicatorBackend::ucc) {
+    auto pg_opts = c10d::ProcessGroupUCC::Options::create();
```

Comment:
Seems like pytorch `ProcessGroupUCC` doesn't have member `::Options`. https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/ProcessGroupUCC.cpp

cc @zasdfgbnm

Reply:
and there are some build issues
```
[1/112] Building CXX object nvfuser/CMakeFiles/nvfuser_codegen.dir/csrc/multidevice/communicator.cpp.o
FAILED: nvfuser/CMakeFiles/nvfuser_codegen.dir/csrc/multidevice/communicator.cpp.o
...
/opt/pytorch/nvfuser/csrc/multidevice/communicator.cpp: In function 'c10::intrusive_ptr<c10d::Backend> nvfuser::createBackend(nvfuser::CommunicatorBackend, c10::intrusive_ptr<c10d::TCPStore>, nvfuser::RankType, int64_t)':
/opt/pytorch/nvfuser/csrc/multidevice/communicator.cpp:99:52: error: 'create' is not a member of 'c10d::Backend::Options'
   99 |     auto pg_opts = c10d::ProcessGroupUCC::Options::create();
      |                         
```

Reply:
In case people want to build nvfuser with the latest pytorch, you also need to modify nvfuser's CMakeLists:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 00df9b8..8cb7b28 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -219,6 +219,11 @@ target_link_libraries(${NVFUSER_CODEGEN} PRIVATE stdc++fs)
 target_link_libraries(${NVFUSER_CODEGEN} PRIVATE ${CUDA_NVRTC_LIB} ${LIBNVTOOLSEXT})
 target_include_directories(${NVFUSER_CODEGEN} PRIVATE ${CUDA_INCLUDE_DIRS})

+add_library(__nvfuser_ucc INTERFACE)
+target_link_libraries(__nvfuser_ucc INTERFACE ucx::ucs ucx::ucp ucc::ucc)
+target_include_directories(__nvfuser_ucc INTERFACE ${UCC_INCLUDE_DIRS})
+target_link_libraries(${NVFUSER_CODEGEN} PRIVATE __nvfuser_ucc)
+
 # TODO: we should guard the include of gloo
 target_include_directories(${NVFUSER_CODEGEN} PRIVATE ${PROJECT_SOURCE_DIR}/third_party/gloo)
 target_include_directories(${NVFUSER_CODEGEN} PUBLIC
```

Reply:
I think this part should be just:
```C++
  if (backend == CommunicatorBackend::ucc) {
    return c10::make_intrusive<::c10d::ProcessGroupUCC>(store, rank, size);
  }
```

Reply:
hmm, isn't `Option` a memeber of `Backend` which `ProcessGroupUCC` inherits?

ref: https://github.com/pytorch/pytorch/blob/4b29829ece227eef08d6887670db52cec2fc69b8/torch/csrc/distributed/c10d/Backend.hpp#L25-L32

Reply:
That's true, but `ProcessGroupUCC`'s ctor is not using that option (which I think also needs to be fixed): https://github.com/pytorch/pytorch/blob/4b29829ece227eef08d6887670db52cec2fc69b8/torch/csrc/distributed/c10d/ProcessGroupUCC.hpp#L152-L156

Reply:
> I think this part should be just:
> 
> ```c++
>   if (backend == CommunicatorBackend::ucc) {
>     return c10::make_intrusive<::c10d::ProcessGroupUCC>(store, rank, size);
>   }
> ```

Correct, sorry about this.
should we merge a patch with this change only ? lmk if you need action on my side